### PR TITLE
Mask R-CNN

### PR DESCRIPTION
**Pull request by JI411**
* Add base classes (BasePredictor and Coal)
* Add InstanceSegmentationCoal based on Coal
* Add Mask R-CNN model based on BasePredictor 
* Add automatically Annotator based on Mask R-CNN model and mask2vgg converter

### CODE REVIEW

File: src/base.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import sys
+from pathlib import Path
+
+import cv2
+import numpy as np
+from numpy.typing import NDArray
+
+FILE = Path(__file__).resolve()
```
Comment (asromahin) [inline src/base.py:8]: удалить

File: src/base.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    def get_fraction(self) -> NDArray:
+        raise NotImplementedError()
+
+    def plot_on(self, img: NDArray) -> NDArray:
+        raise NotImplementedError()
+
+
+class InstanceSegmentationCoal(Coal):
+
+    def __init__(self, contours):
```
Comment (asromahin) [inline src/base.py:30]: contours -> contour

File: src/base.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    @staticmethod
+    def longest_side(box):
+        return max(np.linalg.norm(box[0] - box[1]), np.linalg.norm(box[1] - box[2]))
+
+    def get_fraction(self) -> float:
+        box = cv2.minAreaRect(self.contours)
+        box = cv2.boxPoints(box)
+        return self.longest_side(box)
+
+    def plot_on(self, img: NDArray) -> NDArray:
```
Comment (asromahin) [inline src/base.py:46]: add parameters (color...)

File: src/converters/vgg_to_mask.py [hunk 4]
```diff
@@ -109,7 +109,7 @@ def save_vgg2mask(source_folder: Path, save_folder: Path, vgg_json: Path,
 
 if __name__ == '__main__':
     save_vgg2mask(
-        source_folder=Path.cwd().parent / "few_data",
-        save_folder=Path.cwd().parent / "few_data_processed",
-        vgg_json=Path.cwd().parent / "few_data.json"
+        source_folder=Path.cwd().parents[1] / "few_data",
```
Comment (asromahin) [inline src/converters/vgg_to_mask.py:112]: Исправить пути

File: src/converters/vgg_to_rotated_yolo.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        image_path = dataset_dir / v['filename']
+        image = cv2.imread(str(image_path))
+        height, width = image.shape[:2]
+
+        annotations = ''
+        for r in v["regions"]:
+            x, y = r["shape_attributes"]["all_points_x"], r["shape_attributes"]["all_points_y"]
+            ((xc, yc), (w, h), theta) = cv2.minAreaRect(np.array(list(zip(x, y))))
+            xc, w = xc / width,  w / width
+            yc, h = yc / height, h / height
```
Comment (asromahin) [inline src/converters/vgg_to_rotated_yolo.py:39]: посмотреть репозиторий и найти что делать с длиной и короткой стороной

File: src/instance_segmentation/dataset.py [hunk 6]
```diff
@@ -10,7 +10,7 @@
 import torch.utils.data
 from PIL import Image
 
-from vgg_to_mask import vgg2dict, dict2mask
+from src.converters.vgg_to_mask import vgg2dict, dict2mask
```
Comment (asromahin) [inline src/instance_segmentation/dataset.py:13]: перенести converters в train

Comment (asromahin) [reply src/instance_segmentation/dataset.py:13]: и весь instance_segmentation в train

Comment (asromahin) [reply src/instance_segmentation/dataset.py:13]: всё связанное с трейном

File: src/instance_segmentation/mask_rcnn.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+from pathlib import Path
+from typing import Union, List
+
+import cv2
+import numpy as np
+import torch
+import torchvision.transforms as transforms
+from numpy.typing import NDArray
+
+FILE = Path(__file__).resolve()
```
Comment (asromahin) [inline src/instance_segmentation/mask_rcnn.py:11]: Удалить

File: src/instance_segmentation/mask_rcnn.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            nms_th: float = 0.2,
+            segmentation_th: float = 0.7,
+            device: str = None
+    ):
+        self.device = self.get_device(device=device)
+        self.model = self.get_model(weights, box_conf_th, nms_th)
+        self.segmentation_th = torch.Tensor([segmentation_th])
+        self.segmentation_th.to(self.device)
+
+    @staticmethod
```
Comment (asromahin) [inline src/instance_segmentation/mask_rcnn.py:34]: перенести staticmethods в отдельный скрипт

File: src/instance_segmentation/mask_rcnn_annotator.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+from pathlib import Path
```
Comment (asromahin) [inline src/instance_segmentation/mask_rcnn_annotator.py:1]: to train
